### ISSUE TITLE
Block fetch fails after re-start

### ISSUE THREAD

**Issue by coot** [input-output-hk/ouroboros-network#749]
After restarting `cardano-node` it connects to other peers but it does not receive blocks:
```
[iohk.cardano.node.CoreId 1:Debug:ThreadId 22] [2019-07-11 14:48:46.05 UTC] Updated chain: [(genesis->8a80f267,69786,0,[]),....
[iohk.cardano.node.CoreId 1:Debug:ThreadId 21] [2019-07-11 14:48:46.05 UTC] [TraceLabelPeer (Peer {localAddr = 127.0.0.1:37145, remoteAddr = 127.0.0.1:3002}) (Left FetchDeclineChainNotPlausible)]
[iohk.cardano.node.CoreId 1:Debug:ThreadId 173] [2019-07-11 14:48:47.19 UTC] Recv MsgRequestRangeChainRange (Point {pointSlot = SlotNo {unSlotNo = 69786}, pointHash = BlockHash 8a80f267}) (Point {pointSlot = SlotNo {unSlotNo = 69786}, pointHash = BlockHash 8a80f267})
[iohk.cardano.node.CoreId 1:Debug:ThreadId 173] [2019-07-11 14:48:47.19 UTC] Send MsgNoBlocks
```

The above logs belong to node which was an upstream node to the one that was shut down.  What we see is that the new node is requesting block `8a80f267` which clearly is on the node's chain, and still the node replies with `MsgNoBlocks`.

**Comment by coot** [comment]
Steps to reproduce it:

* run the `cardano-node` `./script/demo.sh`
* after some time, shutdown one of the nodes
* wait and restart it

You should see that the restarted node builds a new chain from the begining, and is not adopting the block chain presented from the other nodes (since it receives only headers of that blockchain).

If one enables tracing in the block fetch protocol, one can observer that the node is sending `MsgNoBlocks`.

**Comment by mrBliss** [comment]
I have just tried to reproduce this in my branch (https://github.com/input-output-hk/cardano-node/pull/52) with `--bft` and `--real-pbft`, but after a restart of the nodes (note that their chain is being restored), I'm seeing  messages like:
```
iohk.cardano.node.CoreId 0:Debug:ThreadId 81] [2019-07-12 10:55:58.05 UTC] TraceLabelPeer (NodeAddress "127.0.0.1" "3002") (CompletedBlockFetch (Point {getPoint = At (Block {blockPointSlot = SlotNo {unSlotNo = 19679}, blockPointHash = AbstractHash ef2e87d811656786c5c2d8787103776edf3a47e852f17531f01d396a39f45f3e})}) (PeerFetchInFlight {peerFetchReqsInFlight = 1, peerFetchBytesInFlight = 0, peerFetchBlocksInFlight = fromList []}) (PeerFetchInFlightLimits {inFlightBytesHighWatermark = 4400, inFlightBytesLowWatermark = 2200}) (PeerFetchStatusReady (fromList [])))
```

**Comment by coot** [comment]
Does the restarted node adopts the presented chain?

**Comment by mrBliss** [comment]
> Does the restarted node adopts the presented chain?

Yes.

The BlockFetch server is using the ChainDB iterators to stream the blocks. It could be that there's a bug in the mock implementation of the ChainDB iterators. So switching to the real ChainDB (iterator) implementation might have fixed it (most times it will be the other way round :slightly_smiling_face: ). I know the mock and the real implementations differ for some edge cases (although this doesn't seem to be such a case).

I just realised that I was restarting *all* nodes, not just one. Will try some more now.

**Comment by coot** [comment]
Cool, fingers crossed :)

**Comment by mrBliss** [comment]
Your branch together with the ChainDB implementation seems to be working fine! I'm killing and restarting nodes here and there, wiping a database or part of it, and they all seem to end up with the same chain again :+1:

**Comment by coot** [comment]
Great, thank you @mrBliss

**Comment by mrBliss** [comment]
It worked perfectly with `--bft`, I still have to test with`--real-pbft`.

**Comment by mrBliss** [comment]
The issue I had with `--real-pbft` was fixed in #752. Closing this issue again.
